module/epochs/epoch_lookup.go

```diff
@@ -1,103 +1,274 @@
 package epochs
 
 import (
-	"errors"
 	"fmt"
+	"sync"
 
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
 	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/state/protocol/events"
 )
 
-// EpochLookup implements the EpochLookup interface using protocol state to
-// match views to epochs.
+// epochRange captures the counter and view range of an epoch (inclusive on both ends)
+type epochRange struct {
+	counter   uint64
+	firstView uint64
+	finalView uint64
+}
+
+func (epoch epochRange) equals(other epochRange) bool {
+	return epoch.counter == other.counter &&
+		epoch.firstView == other.firstView &&
+		epoch.finalView == other.finalView
+}
+
+// epochRangeCache stores at most the 3 latest epoch ranges.
+// Ranges are ordered by counter (ascending) and right-aligned.
+// Not safe for concurrent use.
+type epochRangeCache [3]*epochRange
+
+// latest returns the latest cached epoch range, or nil if no epochs are cached.
+func (cache *epochRangeCache) latest() *epochRange {
+	return cache[2]
+}
+
+// combinedRange returns the endpoints of the combined view range of all cached
+// epochs. In particular, we return the lowest firstView and the greatest finalView.
+// At least one epoch must already be cached, otherwise this function will panic.
+func (cache *epochRangeCache) combinedRange() (firstView uint64, finalView uint64) {
+
+	// low end of the range is the first view of the first cached epoch
+	for _, epoch := range cache {
+		if epoch != nil {
+			firstView = epoch.firstView
+			break
+		}
+	}
+	// high end of the range is the final view of the latest cached epoch
+	finalView = cache.latest().finalView
+	return
+}
+
+// add inserts an epoch range to the cache.
+// Validates that epoch counters and view ranges are sequential.
+// Adding the same epoch multiple times is a no-op.
+// Guarantees ordering and alignment properties of epochRangeCache are preserved.
+// No errors are expected during normal operation.
+func (cache *epochRangeCache) add(epoch epochRange) error {
```

Comment:
any specific reason for accepting value here?

Reply:
https://github.com/onflow/flow-go/pull/2927/commits/7359c32874c26889d27600c49f29b45c7578617e